File: apps/control-api/src/control_api/protocol.py
from __future__ import annotations

import json
import re
import uuid
from datetime import datetime
from typing import Annotated, Any, Literal

from pydantic import BaseModel, BeforeValidator, ConfigDict, Field, field_validator, model_validator

from .security import canonicalize_workspace_roots

PROTOCOL_VERSION = 1
MAX_FRAME_BYTES = 262_144
MAX_ENSURE_ASCII_PARAMS_BYTES = 229_376
MAX_ENSURE_ASCII_PROJECTED_BYTES = 229_376
MAX_ENSURE_ASCII_TEXT_BYTES = 200_000
MAX_TEXT_CODE_POINTS = 200_000
MAX_SEQUENCE = 9_223_372_036_854_775_807
_CANONICAL_UUID = re.compile(
    r"^[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}$"
)
_RFC3339_DATETIME = re.compile(
    r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})$"
)
AllowedRpcMethod = Literal[
    "model/list",
    "thread/start",
    "thread/list",
    "thread/read",
    "thread/resume",
    "turn/start",
    "turn/steer",
    "turn/interrupt",
]
ProtocolErrorCode = Literal[
    "invalid_envelope",
    "unsupported_version",
    "unauthorized",
    "policy_denied",
    "invalid_epoch",
    "expired",
    "conflict",
    "upstream_error",
    "invalid_upstream",
    "indeterminate",
    "internal_error",
]
ALLOWED_RPC_METHODS = frozenset(
    {
        "model/list",
        "thread/start",
        "thread/list",
        "thread/read",
        "thread/resume",
        "turn/start",
        "turn/steer",
        "turn/interrupt",
    }
)

DEVICE_OUTBOUND_TYPES = frozenset(
    {"hello", "heartbeat", "heartbeat_ack", "command_ack", "event", "approval_request", "error"}
)
SERVER_OUTBOUND_TYPES = frozenset(
    {"hello_ack", "heartbeat", "heartbeat_ack", "command", "approval_decision", "token_refresh"}
)


class WireModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


def _require_canonical_uuid(value: Any) -> Any:
    if isinstance(value, uuid.UUID):
        return value
    if not isinstance(value, str) or _CANONICAL_UUID.fullmatch(value) is None:
        raise ValueError("UUID fields must use canonical hyphenated notation")
    return value


def _require_rfc3339_datetime(value: Any) -> Any:
    if isinstance(value, datetime):
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("datetime values must include a UTC offset")
        return value
    if not isinstance(value, str) or _RFC3339_DATETIME.fullmatch(value) is None:
        raise ValueError("datetime fields must use RFC 3339 notation")
    return value


WireUUID = Annotated[uuid.UUID, BeforeValidator(_require_canonical_uuid)]
WireDateTime = Annotated[datetime, BeforeValidator(_require_rfc3339_datetime)]


def _wire_json_default(value: Any) -> str:
    if isinstance(value, uuid.UUID):
        return str(value)
    if isinstance(value, datetime):
        return value.isoformat().replace("+00:00", "Z")
    raise TypeError(f"{type(value).__name__} is not JSON serializable")


def ensure_ascii_json_byte_length(value: Any) -> int:
    try:
        encoded = json.dumps(
            value,
            ensure_ascii=True,
            separators=(",", ":"),
            allow_nan=False,
            default=_wire_json_default,
        )
    except (TypeError, ValueError) as exc:
        raise ValueError("wire value is not JSON serializable") from exc
    return len(encoded.encode("ascii"))


def ensure_ascii_string_content_byte_length(value: str) -> int:
    return ensure_ascii_json_byte_length(value) - 2


class ControlEnvelope(WireModel):
    version: Literal[1]
    id: WireUUID
    device_id: WireUUID
    epoch: str = Field(min_length=16, max_length=128)
    seq: int = Field(strict=True, ge=0, le=MAX_SEQUENCE)
    ack: int | None = Field(default=None, strict=True, ge=0, le=MAX_SEQUENCE)
    type: Literal[
        "hello",
        "hello_ack",
        "heartbeat",
        "heartbeat_ack",
        "command",
        "command_ack",
        "event",
        "approval_request",
        "approval_decision",
        "token_refresh",
        "error",
    ]
    sent_at: WireDateTime
    payload: dict[str, Any]

    @model_validator(mode="before")
    @classmethod
    def validate_frame_budget(cls, value: Any) -> Any:
        if isinstance(value, dict) and ensure_ascii_json_byte_length(value) > MAX_FRAME_BYTES:
            raise ValueError(f"control envelope exceeds the {MAX_FRAME_BYTES}-byte wire budget")
        return value

    @field_validator("version", mode="before")
    @classmethod
    def validate_version_type(cls, value: Any) -> Any:
        if type(value) is not int:
            raise ValueError("control protocol version must be an integer")
        return value

    @model_validator(mode="after")
    def validate_handshake_sequence(self) -> ControlEnvelope:
        is_handshake = self.type in {"hello", "hello_ack"}
        if is_handshake and self.seq != 0:
            raise ValueError("hello and hello_ack must use sequence zero")
        if not is_handshake and self.seq == 0:
            raise ValueError("only hello and hello_ack may use sequence zero")
        WIRE_PAYLOAD_MODELS[self.type].model_validate(self.payload)
        return self


class HelloPayload(WireModel):
    connector_version: str = Field(min_length=1, max_length=64)
    codex_version: str = Field(min_length=1, max_length=64)
    schema_digest: str = Field(min_length=16, max_length=256)
    public_key: str = Field(min_length=40, max_length=256)
    capabilities: list[str] = Field(min_length=1, max_length=32)
    workspace_roots: list[str] = Field(min_length=1, max_length=32)
    resumed_from_seq: int = Field(default=0, strict=True, ge=0, le=MAX_SEQUENCE)

    @field_validator("capabilities")
    @classmethod
    def validate_capabilities(cls, value: list[str]) -> list[str]:
        if len(value) != len(set(value)):
            raise ValueError("capabilities cannot contain duplicates")
        unknown = set(value) - ALLOWED_RPC_METHODS
        if unknown:
            raise ValueError(f"unsupported capabilities: {sorted(unknown)}")
        return value

    @field_validator("workspace_roots")
    @classmethod
    def validate_workspace_roots(cls, value: list[str]) -> list[str]:
        return canonicalize_workspace_roots(value)


class HeartbeatPayload(WireModel):
    status: Literal["ok"]


class HeartbeatAckPayload(WireModel):
    received_seq: int = Field(strict=True, ge=0, le=MAX_SEQUENCE)


class HelloAckPayload(WireModel):
    connection_id: WireUUID
    protocol_version: Literal[1]
    device_id: WireUUID
    epoch: str = Field(min_length=16, max_length=128)
    schema_digest: str = Field(min_length=16, max_length=256)
    received_seq: int = Field(strict=True, ge=0, le=MAX_SEQUENCE)
    heartbeat_timeout_seconds: int = Field(strict=True, ge=5, le=300)

    @field_validator("protocol_version", mode="before")
    @classmethod
    def validate_protocol_version_type(cls, value: Any) -> Any:
        if type(value) is not int:
            raise ValueError("protocol_version must be an integer")
        return value


class CommandPayload(WireModel):
    command_id: WireUUID
    idempotency_key: str = Field(min_length=1, max_length=128)
    method: AllowedRpcMethod
    params: dict[str, Any]
    deadline_at: WireDateTime

    @model_validator(mode="after")
    def validate_param_budget(self) -> CommandPayload:
        if ensure_ascii_json_byte_length(self.params) > MAX_ENSURE_ASCII_PARAMS_BYTES:
            raise ValueError(
                f"command params exceed the {MAX_ENSURE_ASCII_PARAMS_BYTES}-byte "
                "ensure_ascii budget"
            )
        if self.method not in {"turn/start", "turn/steer"}:
            return self
        inputs = self.params.get("input")
        if not isinstance(inputs, list) or len(inputs) != 1 or not isinstance(inputs[0], dict):
            raise ValueError("remote turns require exactly one text input")
        item = inputs[0]
        if set(item) - {"type", "text", "text_elements"} or item.get("type") != "text":
            raise ValueError("remote turns allow text input only")
        if "text_elements" in item and item["text_elements"] != []:
            raise ValueError("remote turn text_elements must be empty")
        text = item.get("text")
        if not isinstance(text, str) or not text.strip() or len(text) > MAX_TEXT_CODE_POINTS:
            raise ValueError(f"turn text must contain 1 to {MAX_TEXT_CODE_POINTS} characters")
        if ensure_ascii_string_content_byte_length(text) > MAX_ENSURE_ASCII_TEXT_BYTES:
            raise ValueError(
                f"turn text exceeds the {MAX_ENSURE_ASCII_TEXT_BYTES}-byte ensure_ascii budget"
            )
        return self


class ProtocolErrorPayload(WireModel):
    code: ProtocolErrorCode
    message: str = Field(min_length=1, max_length=4096)
    retryable: bool = Field(strict=True)

    @field_validator("message")
    @classmethod
    def validate_message_content(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("protocol error message cannot be blank")
        return value


class CommandAckError(ProtocolErrorPayload):
    pass


class CommandAckPayload(WireModel):
    command_id: WireUUID
    state: Literal["accepted", "running", "completed", "failed", "expired", "denied"]
    result: Any = None
    error: CommandAckError | None = None

    @model_validator(mode="after")
    def validate_state_shape(self) -> CommandAckPayload:
        has_result = "result" in self.model_fields_set
        has_error = "error" in self.model_fields_set
        if self.state in {"accepted", "running"} and (has_result or has_error):
            raise ValueError("non-terminal command acknowledgements cannot carry result or error")
        if self.state == "completed" and (not has_result or has_error):
            raise ValueError("completed command acknowledgements require result and forbid error")
        if self.state in {"failed", "expired", "denied"} and (
            not has_error or self.error is None or has_result
        ):
            raise ValueError("failed command acknowledgements require error and forbid result")
        if (
            has_result
            and ensure_ascii_json_byte_length(self.result) > MAX_ENSURE_ASCII_PROJECTED_BYTES
        ):
            raise ValueError(
                f"command result exceeds the {MAX_ENSURE_ASCII_PROJECTED_BYTES}-byte "
                "ensure_ascii budget"
            )
        return self


class DeviceEventPayload(WireModel):
    method: str = Field(min_length=1, max_length=128)
    params: dict[str, Any]

    @model_validator(mode="after")
    def validate_projected_budget(self) -> DeviceEventPayload:
        if ensure_ascii_json_byte_length(self.params) > MAX_ENSURE_ASCII_PROJECTED_BYTES:
            raise ValueError(
                f"event params exceed the {MAX_ENSURE_ASCII_PROJECTED_BYTES}-byte "
                "ensure_ascii budget"
            )
        return self


class ApprovalRequestPayload(WireModel):
    approval_id: WireUUID
    command_id: str = Field(min_length=1, max_length=255)
    kind: Literal["command", "file_change", "permission"]
    summary: str = Field(min_length=1, max_length=512)
    details: dict[str, Any]
    expires_at: WireDateTime


class ApprovalDecisionPayload(WireModel):
    approval_id: WireUUID
    decision: Literal["approve", "deny"]
    decided_at: WireDateTime


class TokenRefreshPayload(WireModel):
    pass


WIRE_PAYLOAD_MODELS: dict[str, type[WireModel]] = {
    "hello": HelloPayload,
    "hello_ack": HelloAckPayload,
    "heartbeat": HeartbeatPayload,
    "heartbeat_ack": HeartbeatAckPayload,
    "command": CommandPayload,
    "command_ack": CommandAckPayload,
    "event": DeviceEventPayload,
    "approval_request": ApprovalRequestPayload,
    "approval_decision": ApprovalDecisionPayload,
    "token_refresh": TokenRefreshPayload,
    "error": ProtocolErrorPayload,
}


def validate_envelope_payload(envelope: ControlEnvelope) -> WireModel:
    """Decode a wire payload with the model selected by the envelope discriminator."""
    return WIRE_PAYLOAD_MODELS[envelope.type].model_validate(envelope.payload)


def require_device_outbound(envelope: ControlEnvelope) -> None:
    if envelope.type not in DEVICE_OUTBOUND_TYPES:
        raise ValueError(f"device cannot send envelope type {envelope.type}")

File: connector/internal/protocol/protocol.go
package protocol

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"path"
	"strings"
	"time"
	"unicode/utf8"
)

const (
	Version                         = 1
	MaxFrameBytes                   = 256 << 10
	MaxProjectedPayloadBytes        = 224 << 10
	MaxSequence              uint64 = 1<<63 - 1
)

const (
	TypeHello            = "hello"
	TypeHelloAck         = "hello_ack"
	TypeHeartbeat        = "heartbeat"
	TypeHeartbeatAck     = "heartbeat_ack"
	TypeCommand          = "command"
	TypeCommandAck       = "command_ack"
	TypeEvent            = "event"
	TypeApprovalRequest  = "approval_request"
	TypeApprovalDecision = "approval_decision"
	TypeTokenRefresh     = "token_refresh"
	TypeError            = "error"
)

var validTypes = map[string]struct{}{
	TypeHello: {}, TypeHelloAck: {}, TypeHeartbeat: {}, TypeHeartbeatAck: {},
	TypeCommand: {}, TypeCommandAck: {}, TypeEvent: {}, TypeApprovalRequest: {},
	TypeApprovalDecision: {}, TypeTokenRefresh: {}, TypeError: {},
}

var validRPCMethods = map[string]struct{}{
	"model/list": {}, "thread/start": {}, "thread/list": {}, "thread/read": {},
	"thread/resume": {}, "turn/start": {}, "turn/steer": {}, "turn/interrupt": {},
}

var validProtocolErrorCodes = map[string]struct{}{
	"invalid_envelope": {}, "unsupported_version": {}, "unauthorized": {},
	"policy_denied": {}, "invalid_epoch": {}, "expired": {}, "conflict": {},
	"upstream_error": {}, "invalid_upstream": {}, "indeterminate": {}, "internal_error": {},
}

var validCommandAckStates = map[string]struct{}{
	"accepted": {}, "running": {}, "completed": {}, "failed": {}, "expired": {}, "denied": {},
}

type Envelope struct {
	Version  int             `json:"version"`
	ID       string          `json:"id"`
	DeviceID string          `json:"device_id"`
	Epoch    string          `json:"epoch"`
	Seq      uint64          `json:"seq"`
	Ack      *uint64         `json:"ack,omitempty"`
	Type     string          `json:"type"`
	SentAt   time.Time       `json:"sent_at"`
	Payload  json.RawMessage `json:"payload"`
}

func (e Envelope) Validate() error {
	if e.Version != Version {
		return fmt.Errorf("unsupported protocol version %d", e.Version)
	}
	if !IsUUID(e.ID) || !IsUUID(e.DeviceID) || utf8.RuneCountInString(e.Epoch) < 16 ||
		utf8.RuneCountInString(e.Epoch) > 128 || e.SentAt.IsZero() {
		return errors.New("envelope is missing a required field")
	}
	if e.Seq > MaxSequence || (e.Ack != nil && *e.Ack > MaxSequence) {
		return errors.New("envelope sequence exceeds the signed BIGINT range")
	}
	if _, ok := validTypes[e.Type]; !ok {
		return fmt.Errorf("unknown envelope type %q", e.Type)
	}
	if len(e.Payload) == 0 || !json.Valid(e.Payload) {
		return errors.New("envelope payload is not valid JSON")
	}
	if e.Type == TypeHello || e.Type == TypeHelloAck {
		if e.Seq != 0 {
			return errors.New("hello handshake envelopes must use sequence zero")
		}
	} else if e.Seq == 0 {
		return errors.New("sequenced envelopes must use a positive sequence")
	}
	return nil
}

// ValidateWire applies the envelope and discriminator-selected payload contract.
// Runtime call sites may retain staged handling by using Validate alone.
func (e Envelope) ValidateWire() error {
	if err := e.Validate(); err != nil {
		return err
	}
	return e.ValidatePayload()
}

// ValidatePayload strictly decodes the payload selected by the envelope type.
func (e Envelope) ValidatePayload() error {
	switch e.Type {
	case TypeHello:
		var payload HelloPayload
		if err := decodeRequiredPayload(e.Payload, &payload,
			"connector_version", "codex_version", "schema_digest", "public_key", "capabilities", "workspace_roots"); err != nil {
			return err
		}
		if payload.ConnectorVersion == "" || utf8.RuneCountInString(payload.ConnectorVersion) > 64 ||
			payload.CodexVersion == "" || utf8.RuneCountInString(payload.CodexVersion) > 64 ||
			utf8.RuneCountInString(payload.SchemaDigest) < 16 || utf8.RuneCountInString(payload.SchemaDigest) > 256 ||
			utf8.RuneCountInString(payload.PublicKey) < 40 || utf8.RuneCountInString(payload.PublicKey) > 256 ||
			len(payload.Capabilities) == 0 || len(payload.Capabilities) > 32 ||
			len(payload.WorkspaceRoots) == 0 || len(payload.WorkspaceRoots) > 32 ||
			payload.ResumedFromSeq > MaxSequence {
			return errors.New("hello payload is incomplete")
		}
		seenCapabilities := make(map[string]struct{}, len(payload.Capabilities))
		for _, method := range payload.Capabilities {
			if _, ok := validRPCMethods[method]; !ok {
				return fmt.Errorf("hello capability %q is not allowed", method)
			}
			if _, duplicate := seenCapabilities[method]; duplicate {
				return fmt.Errorf("hello capability %q is duplicated", method)
			}
			seenCapabilities[method] = struct{}{}
		}
		seenRoots := make(map[string]struct{}, len(payload.WorkspaceRoots))
		for _, root := range payload.WorkspaceRoots {
			if root == "" || len([]byte(root)) > 4096 || !strings.HasPrefix(root, "/") ||
				strings.HasPrefix(root, "//") || root == "/" || path.Clean(root) != root ||
				strings.IndexFunc(root, func(character rune) bool { return character < 0x20 }) >= 0 {
				return errors.New("hello workspace root is invalid")
			}
			if _, duplicate := seenRoots[root]; duplicate {
				return errors.New("hello workspace roots contain a duplicate")
			}
			seenRoots[root] = struct{}{}
		}
		return nil
	case TypeHelloAck:
		var payload HelloAckPayload
		if err := decodeRequiredPayload(e.Payload, &payload,
			"connection_id", "protocol_version", "device_id", "epoch", "schema_digest", "received_seq", "heartbeat_timeout_seconds"); err != nil {
			return err
		}
		if !IsUUID(payload.ConnectionID) || payload.ProtocolVersion != Version || !IsUUID(payload.DeviceID) ||
			utf8.RuneCountInString(payload.Epoch) < 16 || utf8.RuneCountInString(payload.Epoch) > 128 ||
			utf8.RuneCountInString(payload.SchemaDigest) < 16 || utf8.RuneCountInString(payload.SchemaDigest) > 256 ||
			payload.ReceivedSeq > MaxSequence ||
			payload.HeartbeatTimeoutSeconds < 5 || payload.HeartbeatTimeoutSeconds > 300 {
			return errors.New("hello acknowledgement payload is invalid")
		}
		return nil
	case TypeHeartbeat:
		var payload HeartbeatPayload
		if err := decodeRequiredPayload(e.Payload, &payload, "status"); err != nil {
			return err
		}
		if payload.Status != "ok" {
			return errors.New("heartbeat status must be ok")
		}
		return nil
	case TypeHeartbeatAck:
		var payload HeartbeatAckPayload
		if err := decodeRequiredPayload(e.Payload, &payload, "received_seq"); err != nil {
			return err
		}
		if payload.ReceivedSeq > MaxSequence {
			return errors.New("heartbeat acknowledgement sequence exceeds the signed BIGINT range")
		}
		return nil
	case TypeCommand:
		var payload CommandPayload
		if err := decodeRequiredPayload(e.Payload, &payload,
			"command_id", "idempotency_key", "method", "params", "deadline_at"); err != nil {
			return err
		}
		if !IsUUID(payload.CommandID) || payload.IdempotencyKey == "" || utf8.RuneCountInString(payload.IdempotencyKey) > 128 ||
			payload.DeadlineAt.IsZero() {
			return errors.New("command payload is incomplete")
		}
		if _, ok := validRPCMethods[payload.Method]; !ok {
			return fmt.Errorf("command method %q is not allowed", payload.Method)
		}
		var params map[string]any
		if err := json.Unmarshal(payload.Params, &params); err != nil || params == nil {
			return errors.New("command params must be an object")
		}
		return nil
	case TypeCommandAck:
		var payload CommandAckPayload
		if err := decodeRequiredPayload(e.Payload, &payload, "command_id", "state"); err != nil {
			return err
		}
		if !IsUUID(payload.CommandID) {
			return errors.New("command acknowledgement id must be a UUID")
		}
		if _, ok := validCommandAckStates[payload.State]; !ok {
			return fmt.Errorf("unknown command acknowledgement state %q", payload.State)
		}
		fields, err := payloadObject(e.Payload)
		if err != nil {
			return err
		}
		_, hasResult := fields["result"]
		_, hasError := fields["error"]
		switch payload.State {
		case "accepted", "running":
			if hasResult || hasError {
				return errors.New("non-terminal command acknowledgements cannot carry result or error")
			}
		case "completed":
			if !hasResult || hasError {
				return errors.New("completed command acknowledgements require result and forbid error")
			}
		case "failed", "expired", "denied":
			if !hasError || payload.Error == nil || hasResult {
				return errors.New("failed command acknowledgements require error and forbid result")
			}
		}
		return validateProtocolError(payload.Error)
	case TypeEvent:
		var payload EventPayload
		if err := decodeRequiredPayload(e.Payload, &payload, "method", "params"); err != nil {
			return err
		}
		if payload.Method == "" || utf8.RuneCountInString(payload.Method) > 128 || payload.Params == nil {
			return errors.New("event payload is incomplete")
		}
		return nil
	case TypeApprovalRequest:
		var payload ApprovalRequestPayload
		if err := decodeRequiredPayload(e.Payload, &payload,
			"approval_id", "command_id", "kind", "summary", "details", "expires_at"); err != nil {
			return err
		}
		if !IsUUID(payload.ApprovalID) || payload.CommandID == "" || utf8.RuneCountInString(payload.CommandID) > 255 ||
			(payload.Kind != "command" && payload.Kind != "file_change" && payload.Kind != "permission") ||
			payload.Summary == "" || utf8.RuneCountInString(payload.Summary) > 512 || payload.Details == nil || payload.ExpiresAt.IsZero() {
			return errors.New("approval request payload is incomplete")
		}
		return nil
	case TypeApprovalDecision:
		var payload ApprovalDecisionPayload
		if err := decodeRequiredPayload(e.Payload, &payload, "approval_id", "decision", "decided_at"); err != nil {
			return err
		}
		if !IsUUID(payload.ApprovalID) || (payload.Decision != "approve" && payload.Decision != "deny") || payload.DecidedAt.IsZero() {
			return errors.New("approval decision payload is invalid")
		}
		return nil
	case TypeTokenRefresh:
		var payload TokenRefreshPayload
		return decodeRequiredPayload(e.Payload, &payload)
	case TypeError:
		var payload ProtocolError
		if err := decodeRequiredPayload(e.Payload, &payload, "code", "message", "retryable"); err != nil {
			return err
		}
		return validateProtocolError(&payload)
	default:
		return fmt.Errorf("unknown envelope type %q", e.Type)
	}
}

func IsUUID(value string) bool {
	if len(value) != 36 || value[8] != '-' || value[13] != '-' || value[18] != '-' || value[23] != '-' {
		return false
	}
	for index, character := range value {
		if index == 8 || index == 13 || index == 18 || index == 23 {
			continue
		}
		if !((character >= '0' && character <= '9') || (character >= 'a' && character <= 'f') || (character >= 'A' && character <= 'F')) {
			return false
		}
	}
	return true
}

func MarshalPayload(value any) (json.RawMessage, error) {
	data, err := json.Marshal(value)
	return json.RawMessage(data), err
}

func NewID() (string, error) {
	var value [16]byte
	if _, err := rand.Read(value[:]); err != nil {
		return "", err
	}
	value[6] = (value[6] & 0x0f) | 0x40
	value[8] = (value[8] & 0x3f) | 0x80
	encoded := hex.EncodeToString(value[:])
	return encoded[0:8] + "-" + encoded[8:12] + "-" + encoded[12:16] + "-" + encoded[16:20] + "-" + encoded[20:32], nil
}

type HelloPayload struct {
	ConnectorVersion string   `json:"connector_version"`
	CodexVersion     string   `json:"codex_version"`
	SchemaDigest     string   `json:"schema_digest"`
	PublicKey        string   `json:"public_key"`
	Capabilities     []string `json:"capabilities"`
	WorkspaceRoots   []string `json:"workspace_roots"`
	ResumedFromSeq   uint64   `json:"resumed_from_seq"`
}

type HelloAckPayload struct {
	ConnectionID            string `json:"connection_id"`
	ProtocolVersion         int    `json:"protocol_version"`
	DeviceID                string `json:"device_id"`
	Epoch                   string `json:"epoch"`
	SchemaDigest            string `json:"schema_digest"`
	ReceivedSeq             uint64 `json:"received_seq"`
	HeartbeatTimeoutSeconds int    `json:"heartbeat_timeout_seconds"`
}

type HeartbeatPayload struct {
	Status string `json:"status"`
}

type HeartbeatAckPayload struct {
	ReceivedSeq uint64 `json:"received_seq"`
}

type CommandPayload struct {
	CommandID      string          `json:"command_id"`
	IdempotencyKey string          `json:"idempotency_key"`
	Method         string          `json:"method"`
	Params         json.RawMessage `json:"params"`
	DeadlineAt     time.Time       `json:"deadline_at"`
}

type CommandAckPayload struct {
	CommandID string         `json:"command_id"`
	State     string         `json:"state"`
	Result    any            `json:"result,omitempty"`
	Error     *ProtocolError `json:"error,omitempty"`
}

type EventPayload struct {
	Method string         `json:"method"`
	Params map[string]any `json:"params"`
}

type ApprovalRequestPayload struct {
	ApprovalID string         `json:"approval_id"`
	CommandID  string         `json:"command_id"`
	Kind       string         `json:"kind"`
	Summary    string         `json:"summary"`
	Details    map[string]any `json:"details"`
	ExpiresAt  time.Time      `json:"expires_at"`
}

type ApprovalDecisionPayload struct {
	ApprovalID string    `json:"approval_id"`
	Decision   string    `json:"decision"`
	DecidedAt  time.Time `json:"decided_at"`
}

type ProtocolError struct {
	Code      string `json:"code"`
	Message   string `json:"message"`
	Retryable bool   `json:"retryable"`
}

type TokenRefreshPayload struct{}

func validateProtocolError(payload *ProtocolError) error {
	if payload == nil {
		return nil
	}
	if _, ok := validProtocolErrorCodes[payload.Code]; !ok {
		return fmt.Errorf("unknown protocol error code %q", payload.Code)
	}
	if strings.TrimSpace(payload.Message) == "" || utf8.RuneCountInString(payload.Message) > 4096 {
		return errors.New("protocol error message is invalid")
	}
	return nil
}

func decodeRequiredPayload(data []byte, destination any, required ...string) error {
	object, err := payloadObject(data)
	if err != nil {
		return err
	}
	for _, field := range required {
		value, ok := object[field]
		if !ok || bytes.Equal(bytes.TrimSpace(value), []byte("null")) {
			return fmt.Errorf("envelope payload is missing %q", field)
		}
	}
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(destination); err != nil {
		return err
	}
	if err := decoder.Decode(&struct{}{}); !errors.Is(err, io.EOF) {
		if err == nil {
			return errors.New("envelope payload contains trailing JSON")
		}
		return err
	}
	return nil
}

func payloadObject(data []byte) (map[string]json.RawMessage, error) {
	var object map[string]json.RawMessage
	if err := json.Unmarshal(data, &object); err != nil || object == nil {
		return nil, errors.New("envelope payload must be an object")
	}
	return object, nil
}
